### PR TITLE
prov/rxd: Add RMA read-write functionality 

### PR DESCRIPTION
**Pull request by jithinjosepkl**
- prov/rxd: Add RMA read-write functionality
- common/proto: Add message types for RMA operations
- prov/udp: Fix memory leaks in UDP CQ
- prov/rxd: Coverity scan fixes

### CODE REVIEW

**Comment by jithinjosepkl** [comment]
This PR uses util-mr, please do not merge until util-mr PR is merged.

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by lanl-ompi** [comment]
Test FAILed.

File: prov/rxd/src/rxd_cq.c [hunk 1]
```diff
@@ -906,8 +981,69 @@ int rxd_process_start_data(struct rxd_ep *ep, struct rxd_rx_entry *rx_entry,
...
 	case ofi_op_write:
+		rma_iov = (struct ofi_rma_iov *) pkt_start->data;
+		for (i = 0; i < rx_entry->op_hdr.iov_count; i++) {
+			ret = util_mr_verify_key(&ep->domain->util_domain,
+						 rma_iov[i].key, &rma_iov[i].addr,
+						 rma_iov[i].len, FI_WRITE);
+			if (ret) {
+				/* todo: handle invalid key case */
+				FI_WARN(&rxd_prov, FI_LOG_EP_CTRL, "invalid key/access permissions\n");
+				return -FI_ENOMEM;
```
Comment (j-xiong) [inline prov/rxd/src/rxd_cq.c]: FI_EACCESS or FI_EINVAL is more suitable here.

File: prov/rxd/src/rxd_cq.c [hunk 2]
```diff
@@ -906,8 +981,69 @@ int rxd_process_start_data(struct rxd_ep *ep, struct rxd_rx_entry *rx_entry,
...
+		tx_entry->peer = rx_entry->peer;
+		tx_entry->read_rsp.iov_count = rx_entry->op_hdr.iov_count;
+		for (i = 0; i < rx_entry->op_hdr.iov_count; i++) {
+			ret = util_mr_verify_key(&ep->domain->util_domain,
+						 rma_iov[i].key, &rma_iov[i].addr,
+						 rma_iov[i].len, FI_READ);
+			if (ret) {
+				/* todo: handle invalid key case */
+				FI_WARN(&rxd_prov, FI_LOG_EP_CTRL, "invalid key/access permissions\n");
+				return -FI_ENOMEM;
```
Comment (j-xiong) [inline prov/rxd/src/rxd_cq.c]: Same here.

File: prov/rxd/src/rxd_ep.c [hunk 3]
```diff
@@ -660,8 +694,101 @@ int rxd_ep_reply_discard(struct rxd_ep *ep, struct ofi_ctrl_hdr *in_ctrl,
...
+		tx_entry->done = 0;
+		curr_offset = 0;
+		*data_sz = 0;
+		for (i = 0; i < pkt->op.iov_count; i++) {
+			rma_iov.addr = tx_entry->read_req.src_iov[i].addr;
+			rma_iov.len = tx_entry->read_req.src_iov[i].len;
+			rma_iov.key = tx_entry->read_req.src_iov[i].key;
+			memcpy(pkt->data + curr_offset, &rma_iov, sizeof(struct ofi_rma_iov));
+			curr_offset += sizeof(struct ofi_rma_iov);
+			*data_sz += curr_offset;
```
Comment (j-xiong) [inline prov/rxd/src/rxd_ep.c]: This line doesn't look right. Move it out of the loop?

Comment (jithinjosepkl) [inline prov/rxd/src/rxd_ep.c]: Right, it should be outside.

File: prov/rxd/src/rxd_rma.c [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+	memset(&msg, 0, sizeof(msg));
+	msg.msg_iov = iov;
+	msg.desc = desc;
+	msg.iov_count = count;
+	msg.rma_iov_count = 1;
+
+	rma_iov.addr = addr;
+	rma_iov.key = key;
+
+	for (i = 0, len = 0; i < count; i++)
```
Comment (j-xiong) [inline prov/rxd/src/rxd_rma.c]: Check if count is over limit?

Comment (jithinjosepkl) [inline prov/rxd/src/rxd_rma.c]: Yeah, we can have that check in debug build. Will add it

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by jithinjosepkl** [comment]
We can merge this one once the util-mr PR (#2212) is merged.

**Pull request merged**